File: ProjectEuler/816.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

#define N 2000000
#define TRIM 250LL*250
#define MOD 50515093

vector <pair <ll, ll> > s(N);

void fill(){
    s[0].first = 290797;
    s[0].second = s[0].first*s[0].first;
    s[0].second %= MOD;
    for(int i=1; i<N; i++){
        s[i].first = (s[i-1].second * s[i-1].second)%MOD;
        s[i].second = (s[i].first * s[i].first)%MOD;
    }
    sort(s.begin(), s.end());
}

ll dist(pair <ll, ll> a, pair <ll, ll> b){
    return (a.first-b.first)*(a.first-b.first) + (a.second-b.second)*(a.second-b.second);
}

ll solve(vector <pair <ll, ll> > points){
    int sz = points.size();
    
    if(sz == 2) return dist(points[0], points[1]);
    if(sz == 3)
        return min(dist(points[0], points[1]), min(dist(points[1], points[2]), dist(points[0], points[2])));

    int mid = sz/2;
    ll dl = solve(vector<pair<ll, ll> > (points.begin(), points.begin()+mid));
    ll dr = solve(vector<pair<ll, ll> > (points.begin()+mid, points.end()));
    ll d = min(dl, dr);
    d = min(d, TRIM);

    vector <pair <ll, ll> > strip;
    for(int i=0; i<sz; i++){
        if((points[i].first - points[mid].first)*(points[i].first - points[mid].first) < d)
            strip.push_back(make_pair(points[i].second, points[i].first));
    }
    sort(strip.begin(), strip.end());
    for(int i=0; i<strip.size(); i++){
        for(int j=i+1; j<strip.size(); j++){
            if(strip[j].first - strip[i].first >= d) break;
            d = min(d, dist(strip[i], strip[j]));
        }
    }
    return d;
}

int main(){
    fill();
    ll ans = solve(s);
    cout << sqrt(ans) << " " << ans << '\n';
    return 0;
}

File: ProjectEuler/816/python/816.py
# n = 2000000
n = 20000
MOD = 50515093

s = [[0, 0] for i in range(n)]

def fill():
    s[0][0] = 290797
    s[0][1] = (s[0][0]*s[0][0])%MOD
    for i in range(1, n):
        s[i][0] = s[i-1][1]**2
        s[i][0] %= MOD
        s[i][1] = s[i][0]**2
        s[i][1] %= MOD

def dist(a, b):
    return (a[0]-b[0])**2 + (a[1]-b[1])**2

def solve(points):
    m = len(points)
    if m == 2:
        return dist(points[0], points[1])
    if m == 3:
        return min(
            dist(points[0], points[1]), 
            dist(points[1], points[2]), 
            dist(points[0], points[2])
        )
    
    mid = m//2
    dl = solve(points[:mid])
    dr = solve(points[mid:])
    d = min(dl, dr)

    strip = []
    for i in range(m):
        if abs(points[i][0] - points[m//2][0]) < d:
            strip.append(points[i])

    strip = sorted(strip, key=lambda point: point[1])

    for i in range(len(strip)):
        for j in range(i+1, len(strip)):
            if strip[j][1] - strip[i][1] >= d:
                break
            d = min(d, dist(strip[i], strip[j]))
    return d

fill()
s.sort()
ans = solve(s)

print(ans**0.5, ans)